flink-runtime/src/main/java/org/apache/flink/runtime/rpc/resourcemanager/ResourceManagerToTaskExecutorHeartbeatScheduler.java

```diff
@@ -0,0 +1,242 @@
+	}
+
+	/**
+	 * start to schedule next heartbeat
+	 */
+	public void start() {
+		checkState(!closed, "The heartbeat connection is already closed");
+		sendHeartbeatToTaskManagerLater(1, heartbeatTimeout, heartbeatInterval);
+	}
+
+	/**
+	 * Checks if the heartbeat schedule was closed.
+	 *
+	 * @return True if the heartbeat schedule was closed, false otherwise.
+	 */
+	public boolean isClosed() {
+		return closed;
+	}
+
+	/**
+	 * stop to schedule heartbeat
+	 */
+	public void close() {
+		closed = true;
+	}
+
+	/**
+	 * get the heartbeat interval
+	 *
+	 * @return heartbeat interval
+	 */
+	public long getHeartbeatInterval() {
+		return heartbeatInterval;
+	}
+
+	/**
+	 * send a heartbeat attempt to taskManager, receive the slotReport from TaskManager or failed depends on the future result.
+	 *
+	 * @param attempt
+	 * @param timeoutMillis
+	 */
+	private void sendHeartbeatToTaskManager(final int attempt, final long timeoutMillis) {
+		// eager check for closed to avoid some unnecessary work
+		if (closed) {
+			return;
+		}
+		FiniteDuration timeout = new FiniteDuration(timeoutMillis, TimeUnit.MILLISECONDS);
+		Future<SlotReport> heartbeatResponse = taskExecutorGateway.triggerHeartbeatToResourceManager(resourceManagerLeaderSessionID, timeout);
+
+		heartbeatResponse.onSuccess(new OnSuccess<SlotReport>() {
+
+			@Override
+			public void onSuccess(SlotReport result) throws Throwable {
+				if (!isClosed()) {
+					// notify the slotManager and trigger next time heartbeat
+					resourceManager.handleSlotReportFromTaskManager(result);
+					sendHeartbeatToTaskManagerLater(1, heartbeatTimeout, heartbeatInterval);
+				}
+			}
+		}, resourceManager.getRpcService().getExecutionContext());
+
+		// upon failure, retry
+		heartbeatResponse.onFailure(new OnFailure() {
+			@Override
+			public void onFailure(Throwable failure) {
+				if (!isClosed()) {
+					if (attempt == maxAttempt) {
+						log.error("ResourceManager {} fail to keep heartbeat with taskManager {} after {} attempts",
+							resourceManager.getAddress(), taskExecutorAddress, attempt);
+						closed = true;
+						// mark TaskManager as failed after heartbeat interaction attempts failed for many times
+						resourceManager.getSelf().notifyResourceFailure(resourceID);
+					} else {
+						// upon timeout exception, exponential increase timeout
+						if (failure instanceof TimeoutException) {
+							if (log.isDebugEnabled()) {
+								log.debug("ResourceManager {} lost heartbeat in {} ms with taskManager {} at {} attempts ",
+									resourceManager.getAddress(), timeoutMillis, attempt);
+							}
+
+							long retryTimeoutMillis = Math.min(2 * timeoutMillis, maxHeartbeatTimeout);
+							sendHeartbeatToTaskManager(attempt + 1, retryTimeoutMillis);
+						} else {
+							log.error(
+								"ResourceManager {} fail to keep heartbeat with taskManager due to an error",
+								failure);
+						}
+						sendHeartbeatToTaskManagerLater(attempt + 1, timeoutMillis, delayOnError);
```

Comment:
I think the logic here is not correct, since you schedule two heartbeats in case of a `TimeoutException`.